onnxruntime/core/providers/cpu/nn/conv.cc

```diff
@@ -4,9 +4,153 @@
+                    dilations.data(),
+                    pads.data(),
+                    strides.data(),
+                    output_shape.GetDims().data(),
+                    static_cast<size_t>(M / group_),
+                    &WorkingBufferSize);
+
+    auto working_data = WorkingBufferSize > 0 ? alloc->Alloc(sizeof(float) * WorkingBufferSize) : nullptr;
+    BufferUniquePtr working_buffer(working_data, BufferDeleter(alloc));
+
+    MlasConv(&Parameters,
+             Xdata,
+             W->template Data<float>(),
+             B != nullptr ? B->template Data<float>() : nullptr,
+             static_cast<float*>(working_buffer.get()),
+             Ydata);
+  } else {
+    const int64_t input_image_size = input_shape.Size();
+    const int64_t output_image_size = output_shape.Size();
+    const int64_t kernel_size = TensorShape(kernel_shape).Size();
+    const int64_t X_offset = C / group_ * input_image_size;
+    const int64_t Y_offset = Y->Shape().Size() / Y->Shape()[0] / group_;
+    const int64_t W_offset = W->Shape().Size() / group_;
+    const int64_t kernel_dim = C / group_ * kernel_size;
+    const int64_t col_buffer_size = kernel_dim * output_image_size;
+
+    auto col_data = alloc->Alloc(sizeof(float) * col_buffer_size);
+    BufferUniquePtr col_buffer(col_data, BufferDeleter(alloc));
+    float* col_buffer_data = static_cast<float*>(col_buffer.get());
+
+    TensorShape image_shape = X->Shape().Slice(1);
+    std::vector<int64_t> col_buffer_shape{kernel_dim};
+    col_buffer_shape.insert(col_buffer_shape.end(), output_shape.GetDims().begin(),
+                            output_shape.GetDims().end());
+
+    for (int image_id = 0; image_id < N; ++image_id) {
+      for (int group_id = 0; group_id < group_; ++group_id) {
+        math::Im2colNd<float, CPUMathUtil, StorageOrder::NCHW>(
+            Xdata + group_id * X_offset,
+            image_shape.GetDims().data(),
+            col_buffer_shape.data(),
+            C * input_image_size,
+            col_buffer_size,
+            kernel_shape.data(),
+            strides.data(),
+            dilations.data(),
+            pads.data(),
+            static_cast<int>(kernel_shape.size()),
+            col_buffer_data,
+            &CPUMathUtil::Instance());
+        math::Gemm<float, CPUMathUtil>(
+            CblasNoTrans,
+            CblasNoTrans,
+            M / group_,
+            output_image_size,
+            kernel_dim,
+            1,
+            W->template Data<float>() + group_id * W_offset,
+            col_buffer_data,
+            0,
+            Ydata + group_id * Y_offset,
+            &CPUMathUtil::Instance());
+      }
+
+      if (B != nullptr) {
+        auto Ymatrix = EigenMatrixMap<float>(Ydata, output_image_size, M);
+        auto Bvec = ConstEigenVectorMap<float>(B->template Data<float>(), M);
+        Ymatrix.rowwise() += Bvec.transpose();
+      }
+
+      Xdata += X_offset * group_;
+      Ydata += Y_offset * group_;
+    }
+  }
```

Comment:
fuse_activation needs to be called for both paths here (2D/3D and the other cases). I'll remove the fuse_activation from the MLAS path when I check in my changes to use it.

Reply:
Added.